wiki/development/coding-guidelines.md

```diff
@@ -8,27 +8,185 @@ order: 1
 ---
 
 
-## 1. Indentation
+## 1. Naming Conventions
 
-4 spaces for indentation.
+### 1.1. Variable Names
+#### 1.1.1. Global Variable naming
+All global variables must start with the ACE prefix followed by the component, separated by underscores. Global variables may not contain the fnc_ prefix if the value is not callable code.
 
-```c++
-class Something: Or {
-....class Other {
-........foo = "bar";
-....};
+Example: `ACE_Component_myVariableName`
+
+_For ACE this is done automatically through the usage of the GVAR macro family._
+
+#### 1.1.2. Function naming
+All functions shall use ACE and the component name as a prefix, as well as the fnc_ prefix behind the component name.
+
+Example: `PREFIX_COMPONENT_fnc_functionName`
+
+_For ACE this is done automatically through the usage of the PREP macro_
+
+### 1.2. Files & Config
+
+#### 1.2.1. SQF File extension
+Files containing sqf scripts will always have a file name extension of ".sqf".
+
+#### 1.2.2. Header files
+All header files will always have the file name extension of `.hpp`.
+
+#### 1.2.3. Own SQF File
+All functions shall be put in their own .sqf file.
+
+#### 1.2.4. Config elements
+Config files will be split up into different header files, each with the name of the config and be included in the `config.cpp` of the component.
+Example:
+    #include "ACE_Settings.hpp"
+And in `ACE_Settings.hpp`:
+```
+class ACE_Settings {
+    // Content
 };
 ```
 
-### 1.1 Reasoning
+### 1.3. Stringtable
+All text that will be displayed to a user shall be defined in a `stingtable.xml` file for multi-language support.
+
+* There will not be any empty stringtable language values.
+* All stringtables will follow the format as specified by Tabler and the translation guidelines from .
+
+
+## 2. Macro Usage
+
+### 2.1. Module/PBO specific Macro Usage
+The family of `GVAR` macro's define global variable strings or constants for use within a module. Please use these to make sure we follow naming conventions across all modules and also prevent duplicate/overwriting between variables in different modules. The macro family expands as follows, for the example of the module 'balls':
+
+
+| Macros |  Expands to |
+| -------|---------|
+| `GVAR(face)` | ace_balls_face |
+|`QGVAR(face)` | "ace_balls_face" |
+| `EGVAR(balls,face)` | ace_balls_face |
+| `EGVAR(leg,face)` | ace_leg_face |
+| `QEGVAR(leg,face)` | "ace_leg_face" |
+
+There also exists the FUNC family of Macros:
+
+| Macros  |  Expands to |
+| -------|---------|
+|`FUNC(face)` | ace_balls_fnc_face or the call trace wrapper for that function.|
+|`EFUNC(balls,face)` | ace_balls_fnc_face or the call trace wrapper for that function.|
+|`EFUNC(leg,face) `| ace_leg_fnc_face or the call trace wrapper for that function.|
+|`DFUNC(face)` | ace_balls_fnc_face and will ALWAYS be the function global variable.|
+|`DEFUNC(leg,face)` | ace_leg_fnc_face and will ALWAYS be the function global variable.|
```

Comment:
L75-L79, anything after the actual expanded value should be in `()`, eg:
`ace_leg_fnc_face (and will ALWAYS be the function global variable)`